### PR TITLE
Full COPC implementation

### PR DESCRIPTION
**Pull request by Jean-Romain**
This PR proposes a full implementation of COPC. Most of the file I changed were changed respecting Martin's programming style and mimicking the original code structure. New files I created are written using more C++ style and some C++11 idioms, but I tried to write them in C-like style as much as possible to respect Martin's style.

# Reader

LASlib/LASzip have been modified to support spatial queries as well as depth queries in COPC files. It works exactly the same way as with LAX files, and the commands `-inside`, `-inside_circle`, and `-inside_rectangle` are now functional for COPC files. Additionally, two new commands, `-inside_depth [n]` and `-resolution [x]`, have been introduced to query the point cloud with a specified density. Commands like the following should clip the rectangle [x1 y1 x2 y2]  and query only the first and second levels of the octree i.e. a low density version of the point cloud.

```cpp
las2las -i file.copc.laz -inside x1 y1 x2 y2 -inside_depth 2 -o out.laz
```

However, the output is NOT a COPC file (see writer section).

In my implementation, when streaming COPC files, the LAZ chunks always arrive in a spatially ordered manner to maximize the spatial coherence of the stream, even though COPC files are not inherently spatially coherent. Therefore, the previous command also reorders the points in a relatively spatially coherent way.

Users of LASlib can do:

```cpp
LASreadOpener lasreadopener;
lasreadopener.set_copc_stream_ordered_by_depth();  // useful for some application
lasreadopener.set_copc_stream_ordered_spatially(); // default most desirable order
lasreadopener.set_copc_stream_ordered_by_chunk();  // normal stream in file order
// [...]
LASreader* lasreader = lasreadopener.open();
```

# Writer

LASlib/LASzip are not capable of writing COPC files. Thus, commands like `las2las -i file.las -o file.copc.laz` are not possible. Similarly, commands like `las2las -i file.copc.laz -inside_depth 2 -o out.laz` do not produce a COPC file. I won't go into details on that one; this has already been discussed internally. Instead, I've created a new software, `lascopcindex`, capable of merging many LAS/LAZ files into a massive COPC file. The typical use is:

```
lascopcindex -merged -i *.laz -o out.copc.laz -ondisk -progress
```

This software offers several options, including optimizations for TLS with `-tls`, memory optimization with `-ondisk`, and memory optimization for spatially unordered files with `-unordered`. For more details, please refer to the full documentation.

## Benchmarks

Benchmarks of `lascopcindex` were conducted with COPC files containing up to 10 billion points on a not very recent laptop. These tests were primarily focused on ALS coverage data, but also included large TLS and UAV files. The benchmarks were carried out on a Linux Mint 20.3 operating system, running on an Intel Core i7-5600U CPU @ 2.60GHz (dual-core, 4 threads). The system had 11.4 GB of RAM and used an SSD for data storage (unless otherwise specified in comments). The code was compiled using `g++`.

| Name        | LA[S\|Z] | lidar | Files | Pts     | Area     | p/m² | Time    | Memory     | Comment                    |
|-------------|------|-------|-------|---------|----------|------|---------|------------|----------------------------|
| autzen      | LAS  | ALS   | 1     | 10.6 million  | 1.4 km²  | 7    | 0:25    | 650 M      | no spatial coherence       |
| autzen      | LAZ  | ALS   | 1     | 10.6 million  | 1.4 km²  | 7    | 0:36    | 650 M      | no spatial coherence       |
| autzen      | LAS  | ALS   | 1     | 10.6 million  | 1.4 km²  | 7    | 0:40    | 300 M      | -ondisk                    |
| autzen      | LAS  | ALS   | 1     | 10.6 million  | 1.4 km²  | 7    | 0:27    | 350 M      | Reordered by gps_time      |
| autzen      | LAS  | ALS   | 1     | 10.6 million  | 1.4 km²  | 7    | 0:36    | 170 M      | Reordered + -ondisk        |
| Campbell08  | LAS  | ALS   | 6     | 94.8 million  | 15 km²   | 4.3  | 3:50    | 1.3 G      |                            |
| Campbell08  | LAS  | ALS   | 6     | 94.8 million  | 15 km²   | 4.3  | 5:00    | 450 M      | -ondisk                    |
| Campbell04  | LAS  | ALS   | 12    | 95.9 million  | 27 km²   | 3.5  | 5:20    | 1.9 G      |                            |
| Campbell04  | LAS  | ALS   | 12    | 95.9 million  | 27 km²   | 3.5  | 8:00    | 430 M      | -ondisk                    |
| Nome creek  | LAZ  | ALS   | 267   | 95 million    | 235 km²  | 0.4  | 5:40    | 300 M      | -ondisk                    |
| Haliburton  | LAS  | ALS   | 361   | 910 million   | 295 km²  | 3.7  | 50:00   | 2.8 G      | I/O on USB HDD             |
| Haliburton  | LAS  | ALS   | 361   | 910 million   | 295 km²  | 3.7  | 2h30    | 1.3 G      | I/O on USB HDD + -ondisk   |
| BCTS        | LAZ  | ALS   | 62    | 1.14  billion  | 111 km²  | 10   | 1h20    | 3.6 G      |                            |
| BCTS        | LAZ  | ALS   | 62    | 1.14 billion  | 111 km²  | 10   | 1h30    | 1.3 G      | I/O on USB HDD + -ondisk   |
| Mtmorency   | LAS  | ALS   | 824   | 1.14 billion  | 188 km²  | 6.1  | 1h10    | 1.2 G      | I/O on USB HDD + -ondisk   |
| CNF07       | LAZ  | ALS   | 1107  | 10.1 billion  | 1100 km² | 6.1  | 15h     | 1.5 G      | I/O on USB HDD + -ondisk   |
| P0300       | LAZ  | TLS   | 1     | 32.8 million  | 3000 m²  | 10k  | 1:50    | 2.2 G      | no spatial coherence       |
| P0300       | LAZ  | TLS   | 1     | 32.8 million  | 3000 m²  | 10k  | 2:00    | 1.1 G      | -tls                       | 
| P0300       | LAZ  | TLS   | 1     | 32.8 million  | 3000 m²  | 10k  | 3:00    | 340 M      | -tls -ondisk               |
| A3P1        | LAZ  | TLS   | 1     | 48.8 million  | 900 m²   | 54k  | 3:10    | 3.2 G      | no spatial coherence       |
| A3P1        | LAZ  | TLS   | 1     | 48.8 million  | 900 m²   | 54k  | 5:16    | 1.0 G      | -tls                       |
| A3P1        | LAZ  | TLS   | 1     | 48.8 million  | 900 m²   | 54k  | 6:00    | 300 M      | -tls -ondisk               |
| Beauce      | LAS  | UAV   | 1     | 1.94 billion  | 2.7 km²  | 700  | 53:00   | 6.3 G      | I/O on USB HD              |
| Beauce      | LAS  | UAV   | 1     | 1.94 billion  | 2.7 km²  | 700  | 2h40    | 1.1 G      | I/O on USB HDD + -ondisk   |

# Documentation

I created the  comprehensive documentation for `lascopcindex`. Furthermore, I updated several other files to include information about the new `-inside_depth` and `-resolution` flags. I also made modifications to the `Makefiles` and other components to ensure compilation on all platform. It compiles with MSVC without any warnings.

### CODE REVIEW

File: src/lascopcindex.cpp [hunk 1]
```diff
@@ -0,0 +1,1541 @@
...
+        }
+      }
+
+      // convert CRS from GeoTIF to OGC WKT
+      if (lasreader->header.vlr_geo_keys)
+      {
+        char* ogc_wkt = 0;
+        I32 len = 0;
+        geoprojectionconverter.set_projection_from_geo_keys(lasheader->vlr_geo_keys[0].number_of_keys, (GeoProjectionGeoKeys*)lasheader->vlr_geo_key_entries, lasheader->vlr_geo_ascii_params, lasheader->vlr_geo_double_params);
+        if (geoprojectionconverter.get_ogc_wkt_from_projection(len, &ogc_wkt))
```
Comment (rsking) [inline src/lascopcindex.cpp:1130]: Wont this write out the error when it does convert to WKT, rather than when it doesn't?

Comment (Jean-Romain) [reply src/lascopcindex.cpp]: Sure, I fixed it. Thank you for catching it.

**Comment by abellgithub** [comment]
What is meant by "spatial coherence"?

**Comment by aritchie-usgs** [comment]
I wonder if lascopcindex could take advantage of index (.lax) files written by lasindex to speed up building .copc files. In my tests lascopcindex is quite slow (~200x slower on large files) compared to the way that QGIS makes copc from laz (automatically when you add them to a project). **BUT** QGIS does not preserve extra bytes fields so the only way to get them from LAZ is by converting with lascopcindex.

**Comment by Jean-Romain** [comment]
@aritchie-usgs I already thought about it and the answer is no, that would bring no improvement. Building a copc files does not imply any spatial query in the files. Yet I'm surprised by your benchmarks. 200x slower is a big claim. I benchmarked it a lot against PDAL and entwine the lascopcindex implementation was consistently 2 to 3 times faster. Sometime, more than that. QGIS uses PDAL.

**Pull request merged**